### PR TITLE
Allow user to configure Rancher Desktop to start upon login on macOS

### LINKED ISSUE #3771: Auto-start option on macOS [2d]
The easiest way to autostart Rancher Desktop at login on macOS is probably via LaunchAgent.

Create the agent file to enable auto-start, and delete it again to disable it:

```console
$ cat ~/Library/LaunchAgents/io.rancherdesktop.autostart.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>io.rancherdesktop.autostart</string>
    <key>ProgramArguments</key>
    <array>
        <string>/Applications/Rancher Desktop.app/Contents/MacOS/Rancher Desktop</string>
    </array>
    <key>ProcessType</key>
    <string>Interactive</string>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <false/>
</dict>
</plist>
```

Update the factory reset code to remove this file!

# Acceptance criteria

* same as 1st entry from #3769

* Enable auto-start, run factory-reset, and confirm that the LaunchAgent file has been deleted.

### PR DESCRIPTION
**Pull request by adamkpickering**
Closes #3771.

### CODE REVIEW

File: src/go/rdctl/pkg/autostart/autostart_darwin.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    <true/>
+    <key>KeepAlive</key>
+    <false/>
+</dict>
+</plist>
+`
+
+var launchAgentFilePath string
+
+func init() {
```
Comment (jandubois) [inline src/go/rdctl/pkg/autostart/autostart_darwin.go:31]: I'm not a big fan of `init` functions because you can't properly do error reporting.

Any reason this needs to be a global variable? You could just inline the call into the code that uses it.

Comment (jandubois) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go]: There is an idiom using `sync.Once` to get the effect of `init`, but defer the initialization until the first time you use the value. See https://github.com/lima-vm/lima/blob/21c7e311ba41aa5d0be537bdc98cf0c3bb2c8769/pkg/osutil/machineid.go#L17-L37 for an example.

But for cheap functions, I think it is best not to bother with caching at all. It only becomes worth it if the function is called a million times. 😄

**Comment by adamkpickering** [comment]
Note that this PR needs a thorough test on macOS, since I don't have a mac to test on.

File: background.ts [hunk 2]
```diff
@@ -165,6 +165,7 @@ Electron.app.whenReady().then(async() => {
       }
     }
     pathManager = getPathManagerFor(cfg.application.pathManagementStrategy);
+    await integrationManager.enforce();
```
Comment (mook-as) [inline background.ts:168]: Does it make sense to do this before applying updates? (I don't see the harm, just curious why the change is being done.)

File: src/go/rdctl/pkg/autostart/autostart_darwin.go [hunk 3]
```diff
@@ -1,9 +1,50 @@
...
 
+const launchAgentFileContents = `<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+    <key>Label</key>
+    <string>io.rancherdesktop.autostart</string>
+    <key>ProgramArguments</key>
+    <array>
+        <string>/Applications/Rancher Desktop.app/Contents/MacOS/Rancher Desktop</string>
```
Comment (mook-as) [inline src/go/rdctl/pkg/autostart/autostart_darwin.go:17]: This will need to actually refer to the application directory.
(I dumped the app on the desktop to try it out, and it totally didn't work…)

**Review by mook-as — CHANGES_REQUESTED** [review]

File: src/go/rdctl/pkg/autostart/autostart_darwin.go (hunk 3 above)
Comment (jandubois) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go:17]: I wonder if we should re-write the auto-start settings every time the app is launched. That way it will fix itself if the user moves the app around.

It is still possible for the user to move the app and log out, and on the next login, the auto-start is broken. But once the user launches the app manually, the auto-start script would be fixed automatically.

This should be a separate PR and not hold up the current work; it can go into the Later milestone.

File: background.ts (hunk 2 above)
Comment (adamkpickering) [reply background.ts:168]: Hmm, I did this when I was moving fast and now that I look into it more deeply, I think it's unnecessary for macOS. It's necessary for Linux though (see [this code](https://github.com/rancher-sandbox/rancher-desktop/pull/3842/files#diff-6d6040c13653d94c34aec6a5c8319bbf0a74124c2c3ae1a2a1d16dc641545248R101)), which is why I was confused.

File: src/go/rdctl/pkg/autostart/autostart_darwin.go (hunk 3 above)
Comment (adamkpickering) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go:17]: Sounds reasonable; I'll make an issue for that.

Comment (mook-as) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go]: Please confirm — this currently won't work at all for me (even without moving the app after starting). If we really want to push even that off to later, I guess I don't care too much, but I wanted to make sure we're explicit on _what_ we're putting into a separate PR.

Comment (jandubois) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go]: > If we really want to push even that off to later

I don't think we should push this part to Later. Users without admin rights will normally move the app to `$HOME/Applications/Rancher Desktop`, so we do need to adjust the launchctl script to the install location.

Moving the app can be addressed later.

**Review by mook-as — COMMENTED** [review]
The file is written correctly-looking, but the app still isn't starting up. (M1, Big Sur, just dropping it on the desktop.)

Still debugging, but I see:
```
Feb 17 15:57:01 hostname Rancher Desktop Helper (GPU)[2758]: getattrlist failed for /System/Library/Frameworks/OpenGL.framework/Resources//GLRendererFloat.bundle/GLRendererFloat: #2: No such file or directory
Feb 17 15:59:38 hostname com.apple.xpc.launchd[1] (io.rancherdesktop.autostart[3701]): Service could not initialize: 20G1120: xpcproxy + 23576 [839][C0C3B45E-76EE-3F7D-9322-9D92BB8CF9B0]: 0xd
Feb 17 15:59:38 hostname com.apple.xpc.launchd[1] (io.rancherdesktop.autostart[3701]): Service exited with abnormal code: 78
```

File: src/go/rdctl/pkg/autostart/autostart_darwin.go [hunk 4]
```diff
@@ -1,9 +1,109 @@
...
 
+const launchAgentFileTemplateContents = `<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+    <key>Label</key>
+    <string>io.rancherdesktop.autostart</string>
+    <key>ProgramArguments</key>
+    <array>
+        <string>{{ .RancherDesktopPath }}</string>
```
Comment (mook-as) [inline src/go/rdctl/pkg/autostart/autostart_darwin.go:23]: This doesn't seem to work by itself. It looks like a workaround can be [found here](https://www.launchd.info/#/Automatically_restart_application); that is, run it via `/usr/bin/open <appdir>` (the `-W` is unnecessary for us, as there's no need to wait before we start).

**Review by mook-as — CHANGES_REQUESTED** [review]
Should `rdctl factory-reset` also delete our autostart file?

File: src/go/rdctl/pkg/autostart/autostart_darwin.go (hunk 4 above)
Comment (jandubois) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go:23]: I'll try this myself once we have a working CI build. The example launchctl script I provided in #3771 was working for me, so not sure what is different now.

**Comment by jandubois** [comment]
> Should rdctl factory-reset also delete our autostart file?

Yes, it is supposed to; see acceptance criteria on #3771

**Comment by adamkpickering** [comment]
> Should `rdctl factory-reset` also delete our autostart file?

Does it not delete the autostart file? There is code to do this in `pkg/factoryreset/delete_data.go`... I'm not sure why it wouldn't be doing this.

File: src/go/rdctl/pkg/autostart/autostart_darwin.go (hunk 4 above)
Comment (jandubois) [reply src/go/rdctl/pkg/autostart/autostart_darwin.go:23]: The problem is that you are just writing

```xml
    <array>
        <string>/Applications/Rancher Desktop.app</string>
    </array>

```

The working script from #3771 provided the full path to the executable, not just the directory:

```xml
    <array>
        <string>/Applications/Rancher Desktop.app/Contents/MacOS/Rancher Desktop</string>
    </array>
```

An alternative is to use `/usr/bin/open -a`:

```xml
    <array>
        <string>/usr/bin/open</string>
        <string>-a</string>
        <string>/Applications/Rancher Desktop.app</string>
    </array>
```

Both work fine. Maybe the latter is preferable because it doesn't need any information about the filenames inside the app directory.

**Comment by mook-as** [comment]
> > Should `rdctl factory-reset` also delete our autostart file?
> 
> Does it not delete the autostart file? There is code to do this in `pkg/factoryreset/delete_data.go`... I'm not sure why it wouldn't be doing this.

Huh, it did end up deleting the file. That's resolved, then.

**Review by mook-as — APPROVED** [review]
LGTM, confirmed that it does autostart for me as appropriate.
